### PR TITLE
50 - Add Args class to serialize arguments into bytes (u32, string)

### PR DESCRIPTION
**Pull request by Thykof**

### CODE REVIEW

File: src/utils/arguments.ts [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+
+    let value = BigInt(0);
+
+    // encoding is little endian
+    for (let i = 3; i >= 1; --i) {
+      value = (value | BigInt(byteArray[this.offset + i])) << BigInt(8);
+    }
+
+    value = value | BigInt(byteArray[this.offset]);
+    this.offset += 4;
```
Comment (gregLibert) [inline src/utils/arguments.ts:64]: You should remove the magic value here. My proposal :

```suggestion
   numBytes = sizeof<u32>();
    if (byteArray.length - this.offset < numBytes) {
      return BigInt(NaN);
    }

    let value = BigInt(0);

    // encoding is little endian
    for (let i = NumBytes-1; i > 0; --i) {
      value = (value | BigInt(byteArray[this.offset + i])) << BigInt(8);
    }

    value = value | BigInt(byteArray[this.offset]);
    this.offset += numBytes;
```

File: src/utils/arguments.ts [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+
+  // Setter
+
+  /**
+   *
+   * @param {BigInt} bigInt
+   * @return {Args}
+   */
+  addU32(bigInt): Args {
+    const u32 = BigInt.asUintN(32, bigInt);
```
Comment (gregLibert) [inline src/utils/arguments.ts:77]:
```suggestion
    const u32 = BigInt.asUintN(/* num of significant bits */ 32, bigInt);
```

File: src/utils/arguments.ts [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+    const byteArray = new Uint8Array(4);
+    for (let i = 0; i < byteArray.length; i++) {
+      // extracts the value of the first, second, third or forth byte by moving
+      // the _select all bits_ (0xFF) mask
+      const targetByte = u32 & (BigInt(0xff) << BigInt(8 * i));
+      // reduces the extracted value to one byte by removing
+      // already encoded and zeroed byte
+      const byte = targetByte >> BigInt(8 * i);
+      byteArray[i] = Number(byte);
+    }
```
Comment (gregLibert) [inline src/utils/arguments.ts:87]:
```suggestion
    const byteArray = new Uint8Array(4);
    const byteMask = BigInt(0xff); // extracts all the bits of a given byte
    for (let i = 0; i < byteArray.length; i++) {
      // extracts the value of the first, second, third or forth byte by moving the mask
      const targetByte = u32 & (byteMask << BigInt(8 * i));
      // reduces the extracted value to targeted byte by removing lower masked bytes
      const byte = targetByte >> BigInt(8 * i);
      byteArray[i] = Number(byte);
    }
```

File: src/utils/arguments.ts [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+  }
+
+  // Setter
+
+  /**
+   *
+   * @param {BigInt} bigInt
+   * @return {Args}
+   */
+  addU32(bigInt): Args {
```
Comment (gregLibert) [inline src/utils/arguments.ts:76]: No argument type ?

File: src/utils/arguments.ts [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+  /**
+   * Adds an argument to the serialized byte string if the argument is an
+   * instance of a handled type (String of 4294967295 characters maximum)
+   *
+   * @param {string} arg the argument to add
+   *
+   * @return {Args} the modified Arg instance
+   */
+  addString(arg: string): Args {
+    if (typeof arg != "string") {
```
Comment (gregLibert) [inline src/utils/arguments.ts:103]: I guess if this function argument type is string, the typeof is also string no ?

File: src/utils/arguments.ts [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+      console.warn(
+        "Wrong arg type, required a string with max length 4294967295"
+      );
+    } else {
+        this.addU32(BigInt(arg.length));
```
Comment (gregLibert) [inline src/utils/arguments.ts:108]: Why not checking the size of the string ?

File: src/utils/arguments.ts [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+
+  // Utils
+
+  /**
+   * Returns a byte string.
+   *
+   * @param {Uint8Array} byteString
+   * @return {string}
+   */
+  toByteString(byteString: Uint8Array): string {
```
Comment (gregLibert) [inline src/utils/arguments.ts:123]: Why not static ? Private ?

Comment (gregLibert) [inline src/utils/arguments.ts:123]: The argument is not a byte string :-)
```suggestion
  toByteString(bytes: Uint8Array): string {
```

**Review by gregLibert — CHANGES_REQUESTED** [review]
:no_good: The lisibility of this code can be easily improved by:
- removing magic numbers
- clarifying variable names
- adding new variables
- changing function to static/private

File: src/utils/arguments.ts (hunk 4 above)
Comment (Thykof) [reply src/utils/arguments.ts:77]: - when I use BigInt it says me Argument of type 'BigInt' is not assignable to parameter of type 'bigint'.ts(2345) at the line bellow
- when I use bigint it says: Cannot find name 'Bigint'. Did you mean 'BigInt'?ts(2552)


I can do 
```
addU32(bigInt: number): Args {
    const u32 = BigInt.asUintN(32, Number(bigInt));
```

File: src/utils/arguments.ts (hunk 5 above)
Comment (Thykof) [reply src/utils/arguments.ts:103]: you're right!

File: src/utils/arguments.ts (hunk 1 above)
Comment (Thykof) [reply src/utils/arguments.ts]: there is no sizeof in typescript

File: src/utils/arguments.ts (hunk 6 above)
Comment (Thykof) [reply src/utils/arguments.ts:108]: ok let's do this and slice the characters that are too much

File: src/utils/arguments.ts (hunk 7 above)
Comment (Thykof) [reply src/utils/arguments.ts:123]: let's mark it as private

File: src/utils/arguments.ts (hunk 4 above)
Comment (gregLibert) [reply src/utils/arguments.ts:77]: :grin:

File: src/utils/arguments.ts (hunk 7 above)
Comment (gregLibert) [reply src/utils/arguments.ts]: If you don't use this, you should make it private :-)

**Review by gregLibert — APPROVED** [review]
:clap: Very first step.

**Pull request merged**